### PR TITLE
Stop hiding BlockPrograms behind the old Programs interface.

### PR DESCRIPTION
**Pull request by pattyshack**
Note: We still need to preserve some of the old Programs' api until we refactor the lru cache in the block computer.

### CODE REVIEW

File: fvm/executionParameters.go [hunk 1]
```diff
@@ -34,7 +36,23 @@ func getEnvironmentMeterParameters(
...
-	env := NewScriptEnvironment(context.Background(), ctx, vm, sth, programs)
+	txnProgs, err := blockProgs.NewSnapshotReadTransactionPrograms(
+		snapshotTime,
+		executionTime)
+	if err != nil {
+		return params, err
+	}
+
+	defer func() {
+		commitErr := txnProgs.Commit()
```
Comment (Tonix517) [inline fvm/executionParameters.go:47]: q: do we also need the `ShouldCommitPrograms()` check here?

File: fvm/fvm.go [hunk 2]
```diff
@@ -88,7 +125,23 @@ func (vm *VirtualMachine) GetAccount(ctx Context, address flow.Address, v state.
...
-	account, err := getAccount(vm, ctx, stTxn, programs, address)
+	txnProgs, err := blockProgs.NewSnapshotReadTransactionPrograms(
+		programs.EndOfBlockExecutionTime,
+		programs.EndOfBlockExecutionTime)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		commitErr := txnProgs.Commit()
```
Comment (Tonix517) [inline fvm/fvm.go:136]: same question above

File: fvm/script.go [hunk 3]
```diff
@@ -113,6 +113,22 @@ func (proc *ScriptProcedure) ShouldDisableMemoryAndInteractionLimits(
 	return ctx.DisableMemoryAndInteractionLimits
 }
 
+func (proc *ScriptProcedure) IsSnapshotReadTransaction() bool {
+	return true
+}
+
+func (proc *ScriptProcedure) ShouldCommitPrograms() bool {
+	return true
```
Comment (Tonix517) [inline fvm/script.go:121]: q: iirc Script is read-only? do we commit programs for Scripts for meters?

File: fvm/executionParameters.go (hunk 1 above)
Comment (pattyshack) [reply fvm/executionParameters.go:47]: I've added ShouldCommitPrograms to deal with bootstrapping.  I'll rename `ShouldCommitPrograms` to `IsBootstrapping` to make it less confusing.

File: fvm/script.go (hunk 3 above)
Comment (pattyshack) [reply fvm/script.go:121]: scripts are read-only and don't modify source code.

Given the same source code, the parsed program will be identical regardless whether it was used by a script or a transaction.  Hence it's safe to commit the program here.

We're committing the program to speed up future transaction/script which uses the same program

File: fvm/fvm.go (hunk 2 above)
Comment (pattyshack) [reply fvm/fvm.go:136]: see above response

**Review by Tonix517 — APPROVED** [review]

**Review by SaveTheRbtz — APPROVED** [review]

File: fvm/programs/programs.go [hunk 4]
```diff
@@ -1,105 +1,29 @@
...
 
-func (p *Programs) Cleanup(modifiedSets ModifiedSets) {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
-	p.currentTxn.AddInvalidator(modifiedSets)
-
-	var err error
-	err = p.currentTxn.Commit()
+// DEPRECATED.  DO NOT USE
```
Comment (pattyshack) [inline fvm/programs/programs.go:19]: NOTE: I had to add this crap to deal with emulator circular dependency

**Comment by pattyshack** [comment]
urgh ...  this needs major refactoring to deal with emulator circular dependency bullshit.  will close for now and tackle the refactoring piecemeal ...

**Pull request closed without merging**